[KFSPTS-4808] Level3 PCard Type 50 record implementation

kyleoliveira opened the pull request:
Implements processing and viewing for the Type 50 record for our US Bank PCard files. More information (including an example upload file) can be found in KFSPTS-4030 (which was the only JIRA for Level3 PCard data originally).

Level3 information is already included in the upload files, this simply adds the ability to read, store, and display it. We're also only dealing with the Type 50 data in this JIRA. After processing a pcard flat file, find a PCDO doc and click on the Transaction Reference Number on its inquiry page. This will show the Procurement Card Transaction Detail Inquiry, which should have the Type 50 (Purchasing Data) information below it.

See the JIRA for the associated SQL file.

If you wish to test this out without tampering with what's in devlocal right now, I can explain how to launch it in the Sandbox environment.

bhutchinson commented on src/main/java/edu/cornell/kfs/fp/batch/service/impl/ProcurementCardCreateDocumentServiceImpl.java:
```diff
@@ -161,6 +167,55 @@ protected ProcurementCardDocument createProcurementCardDocument(List transaction
+          detailExtension = (ProcurementCardTransactionDetailExtendedAttribute) transactionDetail.getExtension();
+      }
+
+      if (ObjectUtils.isNotNull(transaction.getExtension())) {
```
couldn't this be combined with the else above?

kyleoliveira replied:
Yeah, probably.

bhutchinson replied:
yeah, sorry, that was me reading too quickly and thinking the objects being checked for null/not null where the same when they're not; my bad

bhutchinson commented on src/main/java/edu/cornell/kfs/fp/businessobject/PurchasingDataDetail.java:
```diff
@@ -0,0 +1,88 @@
+    private String documentNumber;
+    private Integer financialDocumentTransactionLineNumber;
+
+    public PurchasingDataDetail() {
```
could just remove this and let Java give us a default constructor

kyleoliveira replied:
Ok

bhutchinson commented on src/main/java/edu/cornell/kfs/fp/businessobject/PurchasingDataRecord.java:
```diff
@@ -0,0 +1,38 @@
+    private static final long serialVersionUID = 1L;
+    private Integer transactionSequenceRowNumber;
+
+    public PurchasingDataRecord()
```
another unnecessary constructor

kyleoliveira replied:
Ok

bhutchinson commented on src/main/java/edu/cornell/kfs/fp/businessobject/USBankAddendumRecord.java:
```diff
@@ -0,0 +1,96 @@
+      this.recordId = recordId;
+    }
+
+    public USBankAddendumRecord() {
```
and again

kyleoliveira replied:
Will clean.

bhutchinson commented on src/main/java/edu/cornell/kfs/fp/businessobject/USBankAddendumRecord.java:
```diff
@@ -0,0 +1,96 @@
+     * @return The trimmed substring
+     * @throws StringIndexOutOfBoundsException May occur when taking the actual substring if the provided bounds are beyond the limits of the superstring
+     */
+    protected String extractNormalizedString(String line, int begin, int end) throws StringIndexOutOfBoundsException {
```
it looks like this same code is in multiple classes; I wonder if it could be extracted into a service or utility class to eliminate the duplication

kyleoliveira replied:
That was the goal with this (move all the record parsing to their own classes and base it off of this) but I got sidetracked somewhere...
I think a utility class would be ideal.

bhutchinson commented on src/main/resources/edu/cornell/kfs/fp/businessobject/datadictionary/PurchasingDataRecord.xml:
```diff
@@ -0,0 +1,295 @@
+    <property name="forceUppercase" value="false" />
+    <property name="maxLength" value="16" />
+    <property name="control">
+      <bean parent="TextControlDefinition" p:size="15"/>
```
do we want a control size smaller than the maxLength of the field?

bhutchinson commented on src/main/resources/edu/cornell/kfs/fp/businessobject/datadictionary/PurchasingDataRecord.xml:
```diff
@@ -0,0 +1,295 @@
+    <property name="forceUppercase" value="false" />
+    <property name="maxLength" value="4" />
+    <property name="control">
+      <bean parent="TextControlDefinition" p:size="15"/>
```
do we want a control size so much larger than the maxLength of the field

kyleoliveira replied:
Good catch. I'll fix it up so they match.

bhutchinson commented on src/main/resources/edu/cornell/kfs/fp/businessobject/datadictionary/PurchasingDataRecord.xml:
```diff
@@ -0,0 +1,295 @@
+    <property name="forceUppercase" value="false" />
+    <property name="maxLength" value="12" />
+    <property name="control">
+      <bean parent="TextControlDefinition" p:size="12"/>
```
a common pattern I have seen in base code is that the control size for text controls is two larger than the maxLength for the field (so it would be 14 in this case); not sure if we've been following that pattern too, or if not, if we want to

kyleoliveira replied:
I'm not opposed. I'll go through and update these accordingly, at least to be consistent with base code.

bhutchinson commented on src/main/java/edu/cornell/kfs/fp/batch/ProcurementCardFlatInputFileType.java:
```diff
@@ -176,7 +178,15 @@ public Object parse(byte[] fileByteContent) throws ParseException {
+                      lineCount = ((ProcurementCardTransactionExtendedAttribute)theTransaction.getExtension()).addAddendumLines(bufferedFileReader, lineCount);
+                    }
+                    
+                    // Add the transaction
```
I'm not a fan of comments like this, I don't feel it's adding any value.

kyleoliveira replied:
I'll remove them. They were mostly for planning/outlining purposes.

kyleoliveira commented:
I've added a util class that should refactor the utilities for parsing the PCard. I'm going to test this out in the SB environment before I squash the commit that contains that work.

Feel free to comment on what I've fixed up so far though.

kyleoliveira commented:
Ok, still works. Please review what's up now.

bhutchinson commented on src/main/java/edu/cornell/kfs/fp/businessobject/USBankRecordFieldUtils.java:
```diff
@@ -0,0 +1,194 @@
+        String day = sub.substring(6,8);
+        theDate = Date.valueOf(year+"-"+month+"-"+day);
+    }
+    catch (Exception e) {
```
can we be more specific here than just Exception?

kyleoliveira replied:
I was keeping it the same as the method it replaced, but we can probably throw it as an IOException. What do you think?

bhutchinson replied:
I think we should be as specific as possible with exceptions, both catching and throwing. So I think if we can use IOException here, that would be preferable to Exception.

kyleoliveira replied:
Yeah, I generally agree with/shoot for that. I'll revise the Exceptions thrown by the methods in this class.

bhutchinson replied:
Great, thanks!

bhutchinson commented on src/main/java/edu/cornell/kfs/fp/businessobject/USBankRecordFieldUtils.java:
```diff
@@ -0,0 +1,194 @@
+   * @return The cycle Date parsed from the trimmed substring
+   * @throws Exception When unable to parse the cycle Date
+   */
+  @SuppressWarnings("deprecation")
```
what is generating a deprecation warning and can we avoid it rather than just suppressing it?

kyleoliveira replied:
The Date#setDate method is deprecated in favor of Calendar.set(Calendar.DAY_OF_MONTH, int date).

I didn't want to mess with it due to scope creep, but I can look into it if you think it should really be handled here.

bhutchinson replied:
If you can make the change in less than an hour, I'd say do it. I know we get a lot of warnings now, but I'd rather pay down that tech debt than add to it. org.kuali.kfs.sys.util.KfsDateUtils.newDate() uses Calendar to construct a java.sql.Date in case that's a helpful reference.

bhutchinson commented on src/main/java/edu/cornell/kfs/fp/businessobject/USBankRecordFieldUtils.java:
```diff
@@ -0,0 +1,194 @@
+ * provide subclasses that assist in parsing, but we won't for now.
+ *
+ */
+public class USBankRecordFieldUtils {
```
Seems like this could be more generic, but I'm fine with leaving it specific for now and making it more generic later if/when that becomes beneficial.

kyleoliveira replied:
Ok good.

kyleoliveira commented:
New version pushed up. I'll test this in SB tomorrow to make sure nothing broke, so do not merge yet, but please let me know if there's anything else that needs reworking.

kyleoliveira commented:
Tested in the SB environment and the changes seem to check out. If you're comfortable with the code as it is now, we should be good for merge. There is SQL to be applied, attached to the JIRA.

bhutchinson commented:
I think this is fine. I'm still somewhat conflicted about the Exception catching and throwing in USBankRecordFieldUtils, but I think that's a larger conversation and I think this is ok for now.

I'll defer to @jhulslander on the merge since this PR is assigned to him.

jhulslander commented:
I think is OK
